### PR TITLE
Create BaseServiceException in gcloud-java-core

### PR DESCRIPTION
**Pull request by ajkannan**
(as suggested in #315's discussion)

### CODE REVIEW

**Comment by ajkannan** [comment]
I was contemplating changing the `Code/ErrorInfo` enum to an enum of `DatastoreException`s (in DatastoreException.java), since the `code/httpStatus` and `retryable` fields are repeated in the class and the `Code/ErrorInfo`.  Would that be preferable?

File: gcloud-java-core/src/main/java/com/google/gcloud/BaseServiceException.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud;
+
+/**
+ * Base service exception.
```
Comment (aozarov) [inline gcloud-java-core/src/main/java/com/google/gcloud/BaseServiceException.java]: Base class for all service exceptions.

File: gcloud-java-core/src/main/java/com/google/gcloud/BaseServiceException.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  /**
+   * Returns the code associated with this exception.
+   */
+  public int code() {
+    return code;
+  }
+
+  public boolean retryable() {
```
Comment (aozarov) [inline gcloud-java-core/src/main/java/com/google/gcloud/BaseServiceException.java]: Returns {@code true} when it is safe to retry the operation that caused this exception.

**Comment by ajkannan** [comment]
@aozarov comments incorporated, made enum accessors package private, and renamed enum.

**Pull request merged**